### ISSUE TITLE
Model Train Stopped Before 1500 Epoch for Tutorial

### ISSUE THREAD

**Issue by SammiLyu** [theislab/regvelo#15]
Hello!

Thank you for creating this amazing package!
I am trying to replicate the tutorial and found out that at reg_vae.train(), the model stops training before the 1500 epoch with this message: 
<img width="919" alt="Image" src="https://github.com/user-attachments/assets/79bb29c9-6e49-47ec-a69f-55291b953114" />
And the perturbation results look slightly different from what is shown in the tutorial:
<img width="612" alt="Image" src="https://github.com/user-attachments/assets/bfa13327-5589-4166-a019-9f0561347d52" />

Any clue what may lead to the model stops training early? Any help is much appreciated!

Thank you,
Sammi

**Comment by WWXkenmo** [comment]
Hey, thanks for interesting our work!

From what I see here is that, the model seems a bit slow on your machine (2.43s/it), may I know now information of your working environment? are you using GPU or not?

You can set `early_stopping = False` if model converge earlier.

Best,
Ken

**Comment by SammiLyu** [comment]
Hi Ken,

Thank you so much for the reply!
I tried again with `early_stopping = False` and although the training ran for the entire 1500 epochs, the perturbation results still looks different. Please see the outputs below. (I am new to GPU so my understanding of the message is that I am using the GPU but please correct me if I am wrong) Thank you again for all the help!!
<img width="930" alt="Image" src="https://github.com/user-attachments/assets/3f0259c0-a99e-4554-a4f1-99ab0ad804a3" />
<img width="882" alt="Image" src="https://github.com/user-attachments/assets/ac38ed90-eb5f-4b69-8c70-41ac2b8b14b6" />

Best,
Sammi

**Comment by WWXkenmo** [comment]
Looks a bit weird to me, it should not take such long time to train on example dataset,

Did you carefully follow the tutorial and set `soft_constraint = False`? could you paste your code to me? thanks!

Best,
Ken

**Comment by SammiLyu** [comment]
Certainly & Thank you! Please see the code below! I followed the tutorial and did not change any arguments. `soft_constraint = False` is also set in `reg_vae = REGVELOVI(adata,W=W.T,regulators = TF_list,soft_constraint = False)`

```
import numpy as np
import pandas as pd
import scanpy as sc
import scvelo as scv
import torch
import regvelo
from regvelo import REGVELOVI,sanity_check,prior_GRN,abundance_test,TFscreening
from velovi import preprocess_data
import cellrank as cr
import mplscience
import anndata
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.spatial.distance import cdist

adata = regvelo.datasets.zebrafish_nc()
prior_net = regvelo.datasets.zebrafish_grn()
TF_list = adata.var_names[adata.var["is_tf"]].tolist()

sc.pp.neighbors(adata, n_neighbors=30, n_pcs=50)
scv.pp.moments(adata)

sc.pl.umap(adata, color='cell_type')

adata = preprocess_data(adata)
adata = prior_GRN(adata,prior_net)

TF_list = set(TF_list).intersection(adata.var_names)
TF_list = list(TF_list)
print("final number of TF: "+str(len(TF_list)))

W = adata.uns["skeleton"].copy()
W = torch.tensor(np.array(W))

sparsity = W.sum() / ((W.sum(1) != 0).sum() * W.shape[0])
print("network sparsity: "+str(np.array(sparsity)))

REGVELOVI.setup_anndata(adata, spliced_layer="Ms", unspliced_layer="Mu")
reg_vae = REGVELOVI(adata,W=W.T,regulators = TF_list,soft_constraint = False)

reg_vae.train(early_stopping = False)

reg_vae.save("rgv_model_earlystoppingF")

adata_target = reg_vae.add_regvelo_outputs_to_adata(adata = adata)

scv.tl.velocity_graph(adata_target)

terminal_states = [
        "mNC_head_mesenchymal",
        "mNC_arch2",
        "mNC_hox34",
        "Pigment",
    ]
vk = cr.kernels.VelocityKernel(adata_target)
vk.compute_transition_matrix()
ck = cr.kernels.ConnectivityKernel(adata_target).compute_transition_matrix()
g_raw = cr.estimators.GPCCA(vk)
## evaluate the fate prob on original space
g_raw.compute_macrostates(n_states=8,cluster_key="cell_type")
g_raw.set_terminal_states(terminal_states)
g_raw.compute_fate_probabilities()
g_raw.plot_fate_probabilities(same_plot=False)

g_raw.plot_macrostates(which="terminal", legend_loc="right", s=100)

fate_prob_perturb = []
cand_list = ["nr2f5","sox9b","twist1a","twist1b","sox10","mitfa","tfec","tfap2a","tfap2b"]
for TF in cand_list:
    model = 'rgv_model'
    adata_target_perturb,reg_vae_perturb =  regvelo.in_silico_block_simulation(model,adata,TF)
    ### predict the terminal state dfference
    n_states = 8
    vk = cr.kernels.VelocityKernel(adata_target_perturb)
    vk.compute_transition_matrix()
    ck = cr.kernels.ConnectivityKernel(adata_target_perturb).compute_transition_matrix()
    g = cr.estimators.GPCCA(vk)
    ## evaluate the fate prob on original space
    g.compute_macrostates(n_states=n_states,cluster_key="cell_type")
    g.set_terminal_states(terminal_states)
    g.compute_fate_probabilities()
    ## visualize coefficient
    cond1_df = pd.DataFrame(adata_target_perturb.obsm["lineages_fwd"], columns=adata_target_perturb.obsm["lineages_fwd"].names.tolist())

    fate_prob_perturb.append(cond1_df)

cond2_df = pd.DataFrame(adata_target.obsm["lineages_fwd"], columns=adata_target.obsm["lineages_fwd"].names.tolist())
df_list = []
for i in range(len(fate_prob_perturb)):
    #data = np.array(((fate_prob_perturb[i]+1) - (cond2_df+1)))
    data = abundance_test(cond2_df, fate_prob_perturb[i])
    data = pd.DataFrame({"Score": data.iloc[:,0].tolist(),
                     "Terminal state": data.index.tolist(),
                     "TF":[cand_list[i]]*(data.shape[0])})
    df_list.append(data)

df = pd.concat(df_list)

with mplscience.style_context():
    sns.set(style="whitegrid")
    fig, ax = plt.subplots(figsize=(10,3))
    #sns.barplot(x='Terminal state', y='AUROC',data=data, hue = "Method",palette=pal,ax = ax)
    color_label = "cell_type"
    palette = dict(zip(adata.obs[color_label].cat.categories, adata.uns[f"{color_label}_colors"]))
    subset_palette = {name: color for name, color in palette.items() if name in cond2_df.columns.tolist()}

    sns.barplot(x='TF', y='Score', hue='Terminal state', data=df, ax=ax, palette=palette, dodge=True)

    # Add vertical lines to separate groups
    for i in range(len(df['TF'].unique()) - 1):
        plt.axvline(x=i + 0.5, color='gray', linestyle='--')

    # Label settings
    plt.ylabel('Perturbation coefficient', fontsize=14)
    plt.xlabel('TF', fontsize=14)
    plt.xticks(fontsize=14)  # Increase font size of x-axis tick labels
    plt.yticks(fontsize=14)  # Increase font size of y-axis tick labels
    plt.show()

```

Best,
Sammi

**Comment by WWXkenmo** [comment]
Hey Sammi,

Great, thanks a lot for your patience, after I tried to reproduce, I think I figure it out.

I guess you are using `scvi-tools==1.2.1`, please use scvi-tools lower than 1.2.0 version, I will suggest to use 1.2.0, this is the stable version we used for reproducibility (will come out soon).

I think there has Incompatibility problem between regvelo with the newest scvi-tools version, I will try to figure it out in future.

Please let me know if you still have problem,

All the best,
Ken

**Comment by WWXkenmo** [comment]
Also please use `scvi.settings.seed = 0` for fix the random seed

**Comment by WWXkenmo** [comment]
Hey Sammi,

I asked my colleague, they can reproduce the tutorial perfectly, could you please let me know if problem still exist after you set idea scvi-tools version?

You don't necessary need `early_stopping = False`, just follow tutorial should be fine.

Let me know if I can help anything!

Best,
Ken

**Comment by SammiLyu** [comment]
Hi Ken,

Thank you so much for all the help!
You are right I was using scvi-tools 1.2.1! Changing scvi-tools to 1.2.0 ended up giving me perturbation results very similar to what is in the tutorial (the directions of perturbation all match up, please see below).
<img width="777" alt="Image" src="https://github.com/user-attachments/assets/25700753-1419-4f3c-afa1-082f542e76c6" />
Unlike the tutorial, I still get metric not improving during training if I do not set the early_stopping argument. Do you think this may have something to do with versions of other package as well? (please see below a list of packages in my environment)
<img width="770" alt="Image" src="https://github.com/user-attachments/assets/29fe1203-bf01-4e0e-b149-c7271e7e34ab" />
<img width="343" alt="Image" src="https://github.com/user-attachments/assets/45e3fdc4-f84b-4927-a497-ee36e3b3904f" />
<img width="318" alt="Image" src="https://github.com/user-attachments/assets/8cbda497-6a62-4a8b-866b-684e593370fb" />
<img width="323" alt="Image" src="https://github.com/user-attachments/assets/25ec1d1e-4b19-42ed-af64-82f0dbb0345c" />
<img width="304" alt="Image" src="https://github.com/user-attachments/assets/788e31b3-0b57-4727-8c3b-85f85ccc3f72" />
<img width="281" alt="Image" src="https://github.com/user-attachments/assets/2398c407-297b-44b8-9077-4ed464802f7a" />

**Comment by SammiLyu** [comment]
Hi Ken,

Thank you so much for all the help! Will certainly reach out again if I run into any issue!

Best,
Sammi
